app/vmselect/promql/exec.go

```diff
@@ -13,10 +14,96 @@ import (
 	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
 	"github.com/VictoriaMetrics/metrics"
 	"github.com/VictoriaMetrics/metricsql"
+
+	"github.com/google/uuid"
 )
 
 var logSlowQueryDuration = flag.Duration("search.logSlowQueryDuration", 5*time.Second, "Log queries with execution time exceeding this value. Zero disables slow query logging")
 
+type query struct {
+	q       *string
+	ec      *EvalConfig
+	stopCh  chan error
+	startAt time.Time
+}
+
+type queriesMap struct {
+	mu sync.Mutex
+	m  map[string]query
+}
+
+func newQueriesMap() *queriesMap {
+	var qm queriesMap
+	qm.m = make(map[string]query)
+
+	return &qm
+}
+
+func (qm *queriesMap) Add(q query) string {
+	qm.mu.Lock()
+	c := uuid.New().String()
+	qm.m[c] = q
+	qm.mu.Unlock()
+
+	return c
+}
+
+func (qm *queriesMap) Delete(c string) {
+	qm.mu.Lock()
+	delete(qm.m, c)
+	qm.mu.Unlock()
+}
+
+var runningQueries = newQueriesMap()
+
+const truncateQueryLength = 16
+
+// GetAllRunningQueries get all the running queries' list
+func GetAllRunningQueries() map[string]map[string]string {
+	all := make(map[string]map[string]string)
+	runningQueries.mu.Lock()
+	for c, rq := range runningQueries.m {
+		m := make(map[string]string)
+		if len(*rq.q) > truncateQueryLength {
+			m["query"] = (*rq.q)[:truncateQueryLength] + "..."
+		} else {
+			m["query"] = *rq.q
+		}
+		m["cost"] = time.Since(rq.startAt).String()
+		all[c] = m
+	}
+	runningQueries.mu.Unlock()
+
+	return all
+}
+
+// GetQueryInfo get all the running queries' info
+func GetQueryInfo(c string) (map[string]string, error) {
+	if rq, ok := runningQueries.m[c]; ok {
+		m := make(map[string]string)
+		m["query"] = *rq.q
+		m["start"] = strconv.FormatInt(rq.ec.Start, 10)
+		m["end"] = strconv.FormatInt(rq.ec.End, 10)
+		m["step"] = strconv.FormatInt(rq.ec.Step, 10)
+		m["cost"] = time.Since(rq.startAt).String()
+
+		return m, nil
+	}
+	return nil, fmt.Errorf("query of qid {%v} is not running", c)
+}
+
+// CancelRunningQuery cancel the given query's execution
+func CancelRunningQuery(c string) error {
+	runningQueries.mu.Lock()
+	defer runningQueries.mu.Unlock()
+	if rq, ok := runningQueries.m[c]; ok {
+		rq.stopCh <- fmt.Errorf("cancel query manully")
```

Comment:
While this line ublocks the waiting goroutine on the line 136, it doesn't stop `exec()` goroutine for the "stopped" query. So the goroutine continues consuming compute resources (CPU, RAM, network). So queries aren't stopped - they are just removed from `runningQueries` list.